### ISSUE TITLE
"management-console: ... Failed to migrate the DB... can’t acquire database lock" during upgrade from 3.2.1 to 3.4.1

### ISSUE THREAD

**Issue by dadlerj** [sourcegraph/sourcegraph#4208]
- **Sourcegraph version:** Upgrading from 3.2.1 to 3.4.1
- **Platform information:**

#### Steps to reproduce:

Reported by https://app.hubspot.com/contacts/2762526/company/887065580

>My process:
>
>1. Stop Sourcegraph 3.2.1
>2. Pull and start Sourcegraph 3.4.1
>3. I saw the following error and Sourcegraph failed to startup
>
>`16:06:09 management-console | 2019/05/24 16:06:09 Fatal error connecting to Postgres DB: Failed to migrate the DB. Please contact support@sourcegraph.com for further assistance: timeout: can’t acquire database lock`

#### Expected behavior:

Successful upgrade

#### Actual behavior:

Failed upgrade

**Comment by slimsag** [comment]
Tagging release blocker unless someone has relatively good confidence it isn't.

**Comment by slimsag** [comment]
Actually, we've upgraded two instances from 3.2 without this so probably not a release blocker.

**Comment by mrnugget** [comment]
`frontend` and `management-console` both call this:

https://github.com/sourcegraph/sourcegraph/blob/b7b99e993f89288ab058f51d1eef4d52b6ba6922/pkg/db/dbconn/dbconn.go#L47-L73

That means, they both try to run the migrations at the same time and whoever comes first wins, I guess. That sounds like... a recipe for a bad time.

My proposal:
1. Add a boolean param to this method, `runMigrations bool`, and only frontend passes in `true`, `management-console` does not. If it's `true`, run the migrations.
2. Then I would also add another function that management-console calls: `WaitUntilMigrated` or something like that. It checks whether the database has been migrated yet and if not, it loops until it has been migrated.

What do you think? I keep wondering about the second point, since we don't want to lock out users of the management-console in case a migration is stuck somewhere. But then again: what good is access to the management-console if a migration did not finish?

**Comment by slimsag** [comment]
Running the migrations at the same time and one or the other losing to whoever started first is actually by design (but there could of course be issues here). It's not just the management console and frontend that are competing, it's also _all of the frontend replicas_. e.g. in most customer cluster deployments there are multiple running instances of the frontend and they would also be competing the same way so whatever we do applies to all of them as well.

> I keep wondering about the second point, since we don't want to lock out users of the management-console in case a migration is stuck somewhere. But then again: what good is access to the management-console if a migration did not finish?

I don't think there is any. In this case the DB needs to be fixed, management console is useless.

**Comment by mrnugget** [comment]
> Running the migrations at the same time and one or the other losing to whoever started first is actually by design (but there could of course be issues here). It's not just the management console and frontend that are competing, it's also _all of the frontend replicas_. e.g. in most customer cluster deployments there are multiple running instances of the frontend and they would also be competing the same way so whatever we do applies to all of them as well.

And what was the reasoning behind that design? Redundancy? In that case, wouldn't it be enough if we let the frontend replicas compete?

**Comment by slimsag** [comment]
The only reason the management console does the same thing as the frontend is because both rely on DB access and it makes sense for that logic to be standardized / unified.

The only reason we _wouldn't_ do that is if for some reason competition in migrations was not safe and lead to issues, but because the frontend replicas already compete to perform these migrations there is no risk in also having the management console do so.

In other words, I don't think this issue actually has anything to do with the management console and its interaction with the DB. It could have just as easily happened with two frontend instances and no management console running. Does that make sense?

**Comment by slimsag** [comment]
If your question was around why the frontends compete to perform migrations at all, that is unclear to me but @keegancsmith would probably know IIRC he implemented that logic.

**Comment by mrnugget** [comment]
> In other words, I don't think this issue actually has anything to do with the management console and its interaction with the DB. It could have just as easily happened with two frontend instances and no management console running. Does that make sense?

Yes and you're right.

I have to read up on this again, but the `can't acquire lock` error is something I'd expect if multiple clients tried to migrate the database at the same time.

And if we keep the "multiple clients compete for who-finishes-migrations-first" behavior then at least we should make "losing" clients behave better

**Comment by slimsag** [comment]
Definitely agree. AFAIK the current competing behavior is not well thought out but was deemed functional enough. It should be re-thought from first principles though. IIRC it is also bad in other ways, too, like if migrations take a while due to a lot of data in postgres then other pods/frontend-replicas will be dying constantly due waiting on the container running the migration and this can trigger downtime alerts (something we've seen in prod in the past)

**Comment by keegancsmith** [comment]
I think we should stick to the frontend being the only one responsible for this. We have adjusted ready and health checks to account for migrations for frontend. The frontend won't be killed unexpectedly by kubernetes while migrating. The root cause of this failed migration is likely just that, management-console winning and being killed due to a slower than expected migration.

Additionally we need to make management-console robust to these sort of issues. Management console DB needs are very minor. I would just check if the migrate version is high enough, if so the management-console continues to work.

WDYT @slimsag?

**Comment by keegancsmith** [comment]
Alternatively we can make management console the only pod which is allowed to run migrations? Additionally we make it robust if migrations fail and expose some admin actions to fix migrations (ie mark as not dirty, etc).

**Comment by mrnugget** [comment]
> I think we should stick to the frontend being the only one responsible for this.

And keep the "multiple frontend replicas" race for it behavior?

> The root cause of this failed migration is likely just that, management-console winning and being killed due to a slower than expected migration.

I don't think that the error we're looking at here _is_ a failed migration. I think another process had a lock on the database (because it was migrating it, is my current assumption) and the management-console process tried the same, but timed out.

**Comment by slimsag** [comment]
> Additionally we need to make management-console robust to these sort of issues. Management console DB needs are very minor. I would just check if the migrate version is high enough, if so the management-console continues to work.

Today the management console isn't really useful currently if the DB is screwed up anyway. Even if you can access it and edit/save config, the frontend and all of Sourcegraph would be broken anyway so editing config isn't much good in that situation.

> Alternatively we can make management console the only pod which is allowed to run migrations? Additionally we make it robust if migrations fail and expose some admin actions to fix migrations (ie mark as not dirty, etc).

This sounds super reasonable to me.

"mark as not dirty" could obviously come later the first step here would just be making mgmt console the only one to run migrations. Also not clear to me when an admin _should_ mark the DB as not dirty?

**Comment by keegancsmith** [comment]
everytime we have run into migration issues, it was fixed by marking the DB not dirty and trying again. I don't even think we thought much about the implications of that, its just what we did. Luckily all the migrations I've reviewed/written are written to be idempotent, so this has been fine.

Given that fact, we should probably introduce some sort of bounded retry + testing that migrations are idempotent. However, the latter is hard to do.

**Comment by mrnugget** [comment]
Thanks for the input! I'll try the "management console is the only one running migrations" approach and investigate/read-up on the possibilities regarding mark-as-dirty/retry.

**Comment by mrnugget** [comment]
First update, after I dug into `migrate` a little bit to see how it handles concurrency.

What I found out is great: it uses database locks (in PG's case an advisory lock) to ensure only one process can run migrations. So we _should_ be fine with our multiple processes trying to run the migrations.

Here is the `Up()` method: https://sourcegraph.com/github.com/golang-migrate/migrate@v4.2.3/-/blob/migrate.go#L263

That uses `lock()` and that one uses a timeout when trying to acquire the lock: https://sourcegraph.com/github.com/golang-migrate/migrate@v4.2.3/-/blob/migrate.go#L852

And the error that's produced if the lock is not acquired in the timeout-time is the same as in the issue title: https://sourcegraph.com/github.com/golang-migrate/migrate@v4.2.3/-/blob/migrate.go#L31:46

So, what I thought happened for the customer is that `management-console` and `frontend` were racing each other for the migration and `management-console` couldn't acquire the lock because `frontend` hat it.

Possible fix: increase the "lock timeout" from the default 15 seconds to something like... 5minutes?

The problem is: I can't reproduce this error. I tried this locally with the same version of `migrate` (`v4.2.3`), but both processes just block, no timeout
error:

```
createdb --owner=sourcegraph --encoding=UTF8 testing_migrations
mkdir test_migrations
migrate create -seq -digits 1 -ext sql -dir ./test_migrations no_sleep_till
echo "SELECT pg_sleep(300);" >> test_migrations/1_no_sleep_till.up.sql
```

Then, in one terminal window:

```
migrate -source file://./test_migrations -database \
  'postgres://sourcegraph:sourcegraph@localhost:5432/testing_migrations?application_name=migrate_1' up
```

And in another terminal window:

```
migrate -source file://./test_migrations -database \
  'postgres://sourcegraph:sourcegraph@localhost:5432/testing_migrations?application_name=migrate_2' up
```

I would expect the second process to produce the same error after the timeout, but it doesn't. It just blocks.

I'll keep you updated. But for now, short version: running migrations from multiple processes _should be fine_, we just need to make sure that we're handling the timeouts correctly

**Comment by mrnugget** [comment]
I've got it. It's a race condition.

Before a process tries to migrate, it runs `ensureVersionTable()` in `postgres.WithInstance` which uses the same locking mechanism and the same lock ID.

If both processes run through `postgres.WithInstance` at the same time, before neither could get to the lock in `Up()`, then they will both block on the migration lock and one of the processes will run into a timeout.

But since `ensureVersionTable()` uses the same lock it's hard to replicate this, since you would have to do this:

| t  | Process 1              |           Process 2       |
|----|------------------------|---------------------------|
| 0  |  ensureVersionTable()  |                           |      Time
| 1  |                        |   ensureVersionTable()    |       v
| 2  |   run migrations       |                           |
| 3  |                        |     run migrations        |


Much more likely is it though to run into this:

| t | Process 1              |           Process 2       |
|---|------------------------|---------------------------|
| 0 |  ensureVersionTable()  |                           |      Time
| 1 |   run migrations       |                           |       v
| 2 |                        |   ensureVersionTable()    |
| 3 |                        |     run migrations        |


So, I can't replicate the bug, but I'd say we should increase the `LockTimeout` so that the other processes wait for the migrations and then we'd be fine.

Opinions?

**Comment by keegancsmith** [comment]
Thanks for solving this. Please cherry-pick it into 3.4 and notify @slimsag once you have done that (I believe he is shepherding the 3.4.3 release)

**Comment by mrnugget** [comment]
Done! cherry-picked 148a4c5 into `3.4` and pushed.
